marlowe-apps/proofspace/Network/ProofSpace.hs

```diff
@@ -0,0 +1,272 @@
+  , reqActionId :: String
+  , reqActionInstanceId :: String
+  , reqActionEventId :: String
+  , reqActionParams :: Value
+  , reqReceivedCredentials :: [ReceivedCredential]
+  , reqIssuedProjects :: Value
+  , reqIssuedCredentials :: Value
+  }
+  deriving (Show)
+
+instance FromJSON ProofRequest where
+  parseJSON =
+    withObject "ProofRequest" $
+      \o ->
+        do
+          reqProtocolVersion <- o .: "protocolVersion"
+          reqPublicServiceDid <- o .: "publicServiceDid"
+          reqSubscriberConnectDid <- o .: "subscriberConnectDid"
+          reqActionId <- o .: "actionId"
+          reqActionInstanceId <- o .: "actionInstanceId"
+          reqActionEventId <- o .: "actionEventId"
+          reqActionParams <- o .: "actionParams"
+          reqReceivedCredentials <- o .: "receivedCredentials"
+          reqIssuedProjects <- o .: "issuedProjects"
+          reqIssuedCredentials <- o .: "issuedCredentials"
+          pure ProofRequest{..}
+
+-- | A received ProofSpace credential.
+data ReceivedCredential = ReceivedCredential
+  { schemaId :: String
+  , credentialId :: String
+  , credWalletId :: String
+  , fields :: Value
+  , utcIssuedAt :: Integer
+  }
+  deriving (Show)
+
+instance FromJSON ReceivedCredential where
+  parseJSON =
+    withObject "ReceivedCredential" $
+      \o ->
+        do
+          schemaId <- o .: "schemaId"
+          credentialId <- o .: "credentialId"
+          credWalletId <- o .: "credWalletId"
+          fields <- o .: "fields"
+          utcIssuedAt <- o .: "utcIssuedAt"
+          pure ReceivedCredential{..}
+
+-- | The response for a ProofSpace webhook.
+data ProofResponse = ProofResponse
+  { resServiceDid :: String
+  , resSubscriberConnectDid :: String
+  , resActionEventId :: String
+  , resIssuedCredentials :: [Value]
+  , resRevokedCredentials :: [Value]
+  , resOk :: Bool
+  }
+  deriving (Show)
+
+instance ToJSON ProofResponse where
+  toJSON ProofResponse{..} =
+    object
+      [ "serviceDid" .= resServiceDid
+      , "subscriberConnectDid" .= resSubscriberConnectDid
+      , "actionEventId" .= resActionEventId
+      , "issuedCredentials" .= resIssuedCredentials
+      , "revokedCredentials" .= resRevokedCredentials
+      , "ok" .= resOk
+      ]
+
+-- | Type for processing credentials.
+type ProofHandler = ProofRequest -> Handler ProofResponse
+
+-- | Type for the API.
+type ProofApi = "sync-issue" :> ProofRequestBody :> Post '[JSON] ProofResponse
+
+-- | The API.
+proofApi :: Proxy ProofApi
+proofApi = Proxy
+
+-- | Verify the signature for the request body.
+verifyBody
+  :: Digest
+  -- ^ The digest algorithm.
+  -> SomePublicKey
+  -- ^ The public key.
+  -> BS.ByteString
+  -- ^ The signature, in base-64 encoding.
+  -> BS.ByteString
+  -- ^ The request body.
+  -> Bool
+  -- ^ Whether the signature is valid.
+verifyBody digest pubKey signature body =
+  let signature' = decodeBase64BS signature
+      result = unsafePerformIO $ verifyBS digest signature' pubKey body
```

Comment:
It would be very awkward to eliminate the `unsafePerformIO` because the function inside `contentCheck` does not live in `MonadIO`. It would be possible to put this in `bodyCheck` using `liftBase`, but that would force other awkwardness.